reps/2023-08-18-serve-java-dag-api.md

```diff
@@ -0,0 +1,325 @@
+    return app;
+  }
+}
+
+```
+This code orchestrates an application within a static method named `app`. The CLI command for its deployment is as follows:
+```shell
+$ serve run java:io.ray.serve.repdemo.Text:app
+```
+
+> It should be noted that `java` is mentioned here to represent the programming language used for orchestrating the application. Currently, there is no clear plan on how to expand this parameter in the serve run command. One direct approach that comes to mind is to concatenate it before the `import_path`.
+
+Additionally, similar to the Python `app_builder`, a Java application also supports custom parameters. For example:
+```java
+import io.ray.serve.api.Serve;
+import io.ray.serve.deployment.Application;
+import java.util.Map;
+
+public class Hello {
+
+  public static class HelloWorld {
+    private String message;
+
+    public HelloWorld(String message) {
+      this.message = message;
+    }
+
+    public String call() {
+      return message;
+    }
+  }
+
+  public static Application appBuilder(Map<String, String> args) {
+    return Serve.deployment()
+        .setDeploymentDef(HelloWorld.class.getName())
+        .bind(args.get("message"));
+  }
+}
+
+```
+
+The `appBuilder` method takes a `Map` as input parameter, from which users can retrieve the required `message` parameter. The syntax for deployment is as follows:
+
+
+```shell
+$ serve run java:io.ray.serve.repdemo.Hello:appBuilder message="Hello from CLI"
+```
+
+Furthermore, it is worth mentioning that `appBuilder` also supports user-defined input parameter of custom types, as long as the type includes the specified attributes. For example:
+
+```java
+import io.ray.serve.api.Serve;
+import io.ray.serve.deployment.Application;
+import java.util.Map;
+
+public class Hello {
+
+  public static class HelloWorldArgs {
+    private String message;
+
+    public String getMessage() {
+      return message;
+    }
+
+    public void setMessage(String message) {
+      this.message = message;
+    }
+  }
+
+  public static class HelloWorld {
+    private String message;
+
+    public HelloWorld(String message) {
+      this.message = message;
+    }
+
+    public String call() {
+      return message;
+    }
+  }
+
+  public static Application typedAppBuilder(HelloWorldArgs args) {
+    return Serve.deployment().setDeploymentDef(HelloWorld.class.getName()).bind(args.getMessage());
+  }
+}
+
+```
+
+```shell
+$ serve run java:io.ray.serve.repdemo.Hello:typedAppBuilder message="Hello from CLI"
+```
+
+For the aforementioned `Text.java` file, we can generate the corresponding Serve config file using the `serve build` command:
+
+```shell
+$ serve build io.ray.serve.repdemo.Text:app -o serve_config.yaml
+```
+
+The generated config file looks like this:
+```yaml
+proxy_location: EveryNode
+
+http_options:
+  host: 0.0.0.0
+  port: 8000
+
+grpc_options:
+  port: 9000
+  grpc_servicer_functions: []
+
+applications:
+- name: app
```

Comment:
Note that this can take `args:` parameter as well (will be passed in the same way that it is via command line)

Reply:
Great, the example of adding args in the config file has been supplemented.